tools/clang/test/CodeGenHLSL/dot.hlsl

```diff
@@ -0,0 +1,135 @@
+// RUN: %dxc -T vs_6_0 -DFUNC=dot %s | FileCheck %s
+// RUN: %dxc -T vs_6_0 -DFUNC=mul %s | FileCheck %s
+
+// Verifies correct implementation of dot and mul with vectors for various sizes and types.
+
+// Partially pilfered from SPIRV's intrinsic.dot.hlsl
+
+float4 main(int1 i1[2] : IO, int2 i2[2] : IT, int3 i3[2] : IH, int4 i4[2] : IF,
+            float1 f1[2] : FO, float2 f2[2] : FT, float3 f3[2] : FH, float4 f4[2] : FF,
+            uint1 u1[2] : UO, uint2 u2[2] : UT, uint3 u3[2] : UH, uint4 u4[2] : UF) : SV_Position {
```

Comment:
I think this might have been a bit simpler with separate functions in a library target for each case, but this works too.

Reply:
Actually, perhaps separate tests for IRGen (`-fcgl`), `HLOperationLower`, and SPIR-V gen would have been in keeping with our preferred approach now.

Reply:
The SPIRV generator makes no use of `UMulExtended` and SPIRV doesn't have an explicit mad/fma operator that I can see. I don't think this is applicable to SPIR-V.

I don't mind adding a fcgl variant. I need to do something to prompt the previously glitched tests from running anyway.